### DISCUSSION TITLE
Repeated output messages

### DISCUSSION THREAD

**Discussion by madssakre** [optuna/optuna#5300] [Q&A]
I'm getting the Optuna output messages 18 times in a row. How come? Example below. It is quite annoying when debugging the optuna script because print statements pushed outside the buffer...

[I 2024-03-08 12:22:52,823] Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.
Trial 0 finished with value: 0.3707652408629656 and parameters: {'num_conv_layers': 4, 'num_filter_0': 128.0, 'num_filter_1': 96.0, 'num_filter_2': 80.0, 'num_filter_3': 32.0, 'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 5.0, 'kernel_size_3': 5.0, 'dropout_conv_0': 0.4, 'dropout_conv_1': 0.0, 'dropout_conv_2': 0.4, 'dropout_conv_3': 0.4, 'dropout_fc': 0.0, 'num_neurons': 40, 'optimizer': 'Adam', 'lr': 0.00023270677083837802}. Best is trial 0 with value: 0.3707652408629656.

**Comment by nzw0301** [comment]
Could you share the minimal reproducible code with us?

**Comment by madssakre** [comment]
This is the code. I removed stuff not related to the optuna part.
I wanna say, even though I got the repeated messages, it still ran to the end without errors.

Was run on cpu on mac.

Best

**Reply by nzw0301** [reply]
? I mean a python code.

**Comment by madssakre** [comment]
Like this:


#!/usr/bin/env python3
# -*- coding: utf-8 -*-


import sys
sys.path.append("../scripts/")


import logging

import torch
from torch.utils.data import DataLoader
from list_files import list_files_with_all_regexp
from CustomDatasetClass2 import CustomDataset
from FileManager import copy_script_to_destination, find_available_filename_return_without_extension, make_dir
import torch.nn as nn
import torch.optim as optim
# from tqdm import tqdm
import os
from pathlib import Path
import optuna
from optuna.trial import TrialState
import torch.nn.functional as F
from optuna.samplers import TPESampler
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "max_split_size_mb:32"
n_epochs = 30
Nb = 32
Nc = 1
Noff = 21
NTx = 1
Nt = 20

def clear_console():
    os.system('cls' if os.name == 'nt' else 'clear')



class ModelHam1(nn.Module):
    def __init__(blah blah blah)
        # model
        
def train(model, optimizer, train_dataloader, criterion, device):
    model.train()
    train_loss = 0.0
    for i, batch in enumerate(train_dataloader):   
        
        targets = batch['data']['trT'].squeeze().to(device)
        inputs = batch['data']['trI'].to(device)
        inputs = torch.permute(inputs,(0,2,1))
        outputs = model(inputs)
        targets = targets.view(outputs.size())
        tloss = criterion(outputs, targets)
        optimizer.zero_grad()
        tloss.backward()
        optimizer.step()
    
        train_loss += tloss.item()
    
    average_train_loss = train_loss / len(train_dataloader)
    return average_train_loss

def validate(model, validate_dataloader, criterion, device):
    # Validation loop
    model.eval()
    with torch.inference_mode():
        # if epoch % 2 == 0:
        validate_loss = 0.0
        for i, batch in enumerate(validate_dataloader):
             
             targets = batch['data']['vaT'].squeeze().to(device)
             
             inputs = batch['data']['vaI'].to(device)
             inputs = torch.permute(inputs,(0,2,1))
             outputs = model(inputs)
             targets = targets.view(outputs.size())
             vloss = criterion(outputs, targets)
                
             validate_loss += vloss.item()
        
        average_validate_loss = validate_loss / len(validate_dataloader)
    return average_validate_loss    

def objective(trial):
    num_conv_layers = trial.suggest_int("num_conv_layers", 2,8) 
    
    num_filters = [int(trial.suggest_float("num_filter_"+str(i), 16, 128, step=16))
               for i in range(num_conv_layers)]
    
    kernel_sizes = [int(trial.suggest_float("kernel_size_"+str(i), 3,5, step=2))
               for i in range(num_conv_layers)]
    kernel_sizes_maxpool = 1
    stride_maxpool = 1

    dropout_rates_conv = [trial.suggest_float("dropout_conv_"+str(i), 0.0, 0.4,step=0.2) for i in range(num_conv_layers)]
    
    dropout_rate_fc = trial.suggest_float("dropout_fc", 0.0, 0.4,step=0.2)
    num_neurons = trial.suggest_int("num_neurons", Nt*2, Nt*4, step=Nt)  
    
    
    model = ModelHam1(Nb, Nc, Noff, Nt,trial,num_conv_layers,num_filters,num_neurons,kernel_sizes,kernel_sizes_maxpool,stride_maxpool,num_filters_last_layer,dropout_rates_conv,dropout_rate_fc).to(device)  # Use your Model3 class # ,drop_conv2,drop_fc1

    optimizer_name = trial.suggest_categorical("optimizer", ["Adam"]) # "Adam", "RMSprop", "SGD",
    

    lr = trial.suggest_float("lr", 1e-4, 1e-2, log=True)

    
    optimizer = getattr(optim, optimizer_name)(model.parameters(), lr=lr)

    for epoch in range(n_epochs):
        train(model, optimizer,train_dataloader, criterion, device)  
        loss = validate(model,validate_dataloader, criterion, device)   

        trial.report(loss, epoch)
        if trial.should_prune():
            raise optuna.exceptions.TrialPruned()

    return loss

if __name__ == '__main__':
    clear_console()
    torch.manual_seed(42)
    criterion = nn.MSELoss()
    
    C_Dir = "temp"
    D_Dir = "temp"
    make_dir(D_Dir)
    
    train_files = list_files_with_all_regexp(C_Dir,["train"])
    validate_files = list_files_with_all_regexp(C_Dir,["validate"])
    train_dataset = CustomDataset(train_files)
    validate_dataset = CustomDataset(validate_files)
    
    
    train_dataloader = DataLoader(train_dataset, batch_size=Nb, shuffle=True)
    validate_dataloader = DataLoader(validate_dataset, batch_size=Nb, shuffle=False)

    
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    print('Using device:', device)


    number_of_trials = 100

    sampler = TPESampler(seed=42) 
    
    optuna.logging.get_logger("optuna").addHandler(logging.StreamHandler(sys.stdout))
    study_name = "optuna"  
    study_name = f"{D_Dir}{study_name}"
    study_name, study_number = find_available_filename_return_without_extension(D_Dir,"optuna",".db")
    
    
    storage_name = "sqlite:///{}.db".format(f"{D_Dir}{study_name}")
    study = optuna.create_study(study_name=f"{D_Dir}{study_name}", storage=storage_name,direction="minimize",sampler=sampler)
    
    
    study.optimize(objective, n_trials=number_of_trials, gc_after_trial=True)

    print("\nStudy statistics:")
    print("  Number of finished trials:", len(study.trials))
    print("  Number of pruned trials:", len([t for t in study.trials if t.state == TrialState.PRUNED]))
    print("  Number of complete trials:", len([t for t in study.trials if t.state == TrialState.COMPLETE]))

    best_trial = study.best_trial

    print("\nBest trial:")
    print("  Value:", best_trial.value)
    print("  Params:")
    for key, value in best_trial.params.items():
        print(f"    {key}: {value}")

    df = study.trials_dataframe()
    df.to_csv(f"{D_Dir}{study_name}.csv", index=False)

    print("\nOverall Results (ordered by validate):\n", df)

    most_important_parameters = optuna.importance.get_param_importances(study, target=None)

    print('\nMost important hyperparameters:')
    for key, value in most_important_parameters.items():
        print(f'  {key}:{" "*(15-len(key))}{value*100:.2f}%')

    copy_script_to_destination(__file__, D_Dir,f"_wasrun_with_study_number_{study_number}")

**Reply by nzw0301** [reply]
I'm afraid I cannot get the same messages since the code is not reproducible.

**Reply by madssakre** [reply]
It's ok. 
Out of curiosity, what determines the number of outputs during optimization? On another system (Ubuntu), with a similar Optuna optimization, I'm getting the messages twice:
 

> [I 2024-03-14 09:25:06,736] Trial 2774 finished with value: 0.07691332697868347 and parameters: {'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 3.0, 'kernel_size_3': 5.0, 'num_filter_0': 32.0, 'num_filter_1': 48.0, 'num_filter_2': 48.0, 'num_filter_3': 64.0, 'optimizer': 'SGD', 'lr': 0.08835294881986322}. Best is trial 1824 with value: 0.07688857801258564.
> Trial 2774 finished with value: 0.07691332697868347 and parameters: {'kernel_size_0': 3.0, 'kernel_size_1': 3.0, 'kernel_size_2': 3.0, 'kernel_size_3': 5.0, 'num_filter_0': 32.0, 'num_filter_1': 48.0, 'num_filter_2': 48.0, 'num_filter_3': 64.0, 'optimizer': 'SGD', 'lr': 0.08835294881986322}. Best is trial 1824 with value: 0.07688857801258564.
> [I 2024-03-14 09:25:18,911] Trial 2775 pruned. 
> Trial 2775 pruned. 
> [I 2024-03-14 09:25:31,535] Trial 2776 pruned. 
> Trial 2776 pruned. 
> [I 2024-03-14 09:25:43,597] Trial 2777 pruned. 
> Trial 2777 pruned. 
> [I 2024-03-14 09:25:55,628] Trial 2778 pruned. 
> Trial 2778 pruned. 
> [I 2024-03-14 09:26:08,049] Trial 2779 pruned. 
> Trial 2779 pruned. 
> [I 2024-03-14 09:26:20,056] Trial 2780 pruned. 
> Trial 2780 pruned.

**Reply by nzw0301** [reply]
I think some logging-related things. The reported behaviour seems unusual; I suspect your logging config might be misconfigured.
